lib/archiver.js

```diff
@@ -0,0 +1,168 @@
+var path = require('path')
+var promisify = require('es6-promisify')
+var hyperdrive = require('hyperdrive')
+var datEncoding = require('dat-encoding')
+var discoverySwarm = require('discovery-swarm')
+var swarmDefaults = require('datland-swarm-defaults')
+var mkdirp = require('mkdirp')
+var debug = require('debug')('archiver')
+
+mkdirp = promisify(mkdirp)
+
+// exported api
+// =
+
+module.exports = class Archiver {
+  constructor (config) {
+    this.config = config
+    this.archives = {}
+    this.loadPromises = {}
+  }
+
+  // methods
+  // =
+
+  getArchive (key) {
+    return this.archives[key]
+  }
+
+  isLoadingArchive (key) {
+    return key in this.loadPromises
+  }
+
+  // load archive (wrapper) manages load promises
+  async loadArchive (key) {
+    key = datEncoding.toStr(key)
+
+    // fallback to archive if it exists
+    if (key in this.archives) {
+      return this.archives[key]
+    }
+
+    // fallback to the promise, if it exists
+    if (key in this.loadPromises) {
+      return this.loadPromises[key]
+    }
+
+    // ensure the folder exists
+    var archivePath = this._getArchiveFilesPath(key)
+    await mkdirp(archivePath)
+
+    // run and cache the promise
+    var p = this._loadArchiveInner(archivePath, key)
+    this.loadPromises[key] = p
+
+    // when done, clear the promise
+    const clear = () => delete this.loadPromises[key]
+    p.then(clear, clear)
+
+    // when done, save the archive instance
+    p.then(archive => { this.archives[key] = archive })
+
+    return p
+  }
+
+  async closeArchive (key) {
+    key = datEncoding.toStr(key)
+    var archive = this.archives[key]
+    if (archive) {
+      archive.replicationStreams.forEach(s => s.destroy())
+      archive.swarm.close()
+      await new Promise(resolve => archive.close(resolve))
+      delete this.archives[key]
+    }
+  }
+
+  async closeAllArchives () {
+    return Promise.all(Object.keys(this.archives).map(key =>
+      this.closeArchive(key)
+    ))
+  }
+
+  // internal
+  // =
+
+  _getArchiveFilesPath (key) {
+    return path.join(this.config.dir, 'archives', key.slice(0, 2), key.slice(2))
+  }
+
+  // load archive (inner) main load logic
+  async _loadArchiveInner (archivePath, key) {
+    // create the archive instance
+    var archive = hyperdrive(archivePath, key, {sparse: false})
+    archive.replicationStreams = [] // list of all active replication streams
+
+    // wait for ready
+    await new Promise((resolve, reject) => {
+      archive.ready(err => {
+        if (err) reject(err)
+        else resolve()
+      })
+    })
+
+    // join the swarm
+    this._joinSwarm(archive)
+
+    if (!archive.metadata.length) {
+      // wait to receive a first update
+      await new Promise((resolve, reject) => {
+        archive.metadata.update(err => {
+          if (err) reject(err)
+          else resolve()
+        })
+      })
+    }
+
+    // download the full metadata
```

Comment:
It may not be necessary actually. In Beaker I do it so that I can immediately access the files after I load the archive. I wasnt sure if I'd have a similar need in Hypercloud... probably not, actually. I'll remove it.